pkg/cli/zip.go

```diff
@@ -377,6 +389,94 @@ func maybeAddProfileSuffix(name string) string {
 	return name
 }
 
+type jobTrace struct {
+	jobID   jobspb.JobID
+	traceID tracingpb.TraceID
+}
+
+// dumpTraceableJobTraces collects the traces for some "traceable" jobs that are
+// in a running state. The job types in this list are the ones that have
+// explicitly implemented the TraceableJob interface.
+func (zc *debugZipContext) dumpTraceableJobTraces() error {
+	ctx := context.Background()
+	rows, err := zc.firstNodeSQLConn.Query(ctx,
+		`WITH
+latestprogress AS (
+  SELECT job_id, value
+  FROM system.job_info AS progress
+  WHERE info_key = 'legacy_progress'
+  ORDER BY written desc
+),
+jobpage AS (
+  SELECT id
+  FROM system.jobs@jobs_status_created_idx
+  WHERE (job_type IN ($1, $2, $3, $4)) AND (status IN ($5, $6))
+  ORDER BY id
+)
+SELECT distinct (id), latestprogress.value AS progress
+FROM jobpage AS j
+INNER JOIN latestprogress ON j.id = latestprogress.job_id;`,
+		jobspb.TypeBackup.String(),
+		jobspb.TypeRestore.String(),
+		jobspb.TypeImport.String(),
+		jobspb.TypeReplicationStreamIngestion.String(),
+		"running",
+		"reverting",
+	)
+	defer func() {
+		err := rows.Close()
+		log.Warningf(ctx, "failed to close with error: %v", err)
+	}()
+	if err != nil {
+		return err
+	}
+	vals := make([]driver.Value, 2)
+	jobTraces := make([]jobTrace, 0)
+	for err = rows.Next(vals); err == nil; err = rows.Next(vals) {
+		jobID, ok := vals[0].(int64)
+		if !ok {
+			return errors.New("failed to parse jobID")
+		}
+		progressBytes, ok := vals[1].([]byte)
+		if !ok {
+			return errors.New("failed to parse progress bytes")
+		}
+		progress := &jobspb.Progress{}
+		if err := protoutil.Unmarshal(progressBytes, progress); err != nil {
+			return err
+		}
+		jobTraces = append(jobTraces, jobTrace{jobID: jobspb.JobID(jobID), traceID: progress.TraceID})
+	}
+
+	func() {
+		// Debug zip collection sets this to false since results from the query are
+		// all dumped into txt files. In our case we parse the results of the query
+		// with their respective types and pre-process the information before
+		// dumping into a zip file.
+		zc.firstNodeSQLConn.SetAlwaysInferResultTypes(true)
+		defer zc.firstNodeSQLConn.SetAlwaysInferResultTypes(false)
```

Comment:
I think it's fine based on where it is called, but this assumes that AlwaysInferResultTypes is set to false initially.  Maybe SetAlwaysInferResultTypes should return a func that you can defer that sets it to whatever the previous value was.